Documentation/COMPILATION_STEPS.md

```diff
@@ -0,0 +1,237 @@
+- `ProgramClassBuilder.EndDataDivision` signals the end of the DATA DIVISION, thus the current node will be the program node.
+- `ProgramClassBuilder.OnGobackStatement` signals the encounter of a `GobackStatement`. The method creates the `Goback` node, enters it as a child of the current node (most probably the `ProcedureDivision` node) and exit immediately.
+
+The parsing errors produced by CUP parser are stored in the `Diagnostics` collection of the `TemporarySemanticDocument`.
+
+A secondary grammar file, 'EmptyTypeCobolProgram.cup', is kept up-to-date to provide a starting point for anyone willing to implement their own Cobol parser.
+
+## Step 6 Cross check
+
+|Class|Method|Input|Output|
+|-|-|-|-|
+|`CompilationUnit`|`RefreshProgramClassDocumentSnapshot`|Up-to-date `TemporaryProgramClassDocumentSnapshot`|Updated `ProgramClassDocumentSnapshot`|
+
+The goal of this step is to perform a full visit of the refreshed AST in order to check the semantics of the document. The main goal is to check that all used variables are defined unambiguously but other checks specifically related to some statements are performed as well.
+
+The `CrossCompleteChecker` class is the visitor object that implements those checks. Note that it is configured to skip CodeElements during the visit, as it assumes that CodeElements have already been checked previously. All diagnostics must be created on nodes only during this step because CodeElements persist between compilation cycles whereas nodes are always recreated.
+
+The `BeginNode` method triggers the check for variable definition, while the `VisitXXX` methods implement node-specific checks.
+
+As stated before, the diagnostics are stored inside `Node` instances, they are later collected using a dedicated visitor (see `CompilationUnit.AllDiagnostics` method).
+
+## Step 7 Code analysis
+
+|Class|Method|Input|Output|
+|-|-|-|-|
+|`CompilationUnit`|`RefreshCodeAnalysisDocumentSnapshot`|Up-to-date `ProgramClassDocumentSnapshot`|Updated `CodeAnalysisDocumentSnapshot`|
+
+This final step runs external analyzers on the latest `ProgramClassDocumentSnapshot`. To allow all types of code analysis, the analyzers are given all previous document snapshots (Tokens, ProcessedTokens, CodeElements, TemporaryProgramClass and ProgramClass).
+
+Each analyzer receives all five document snapshot in sequence through calls to their `Inspect` methods. The results of the analysis (mainly diagnostics) are stored in a new instance of `InspectedProgramClassDocument`.
+
+## About snapshots and result consistency.
+
+The parser does not use multi-threading to parallelize source code processing. However the client of the parser may choose to perform multiple operations on the same source simultaneously. Typically in LanguageServer, a thread may be reading results of an incremental compilation round while the timer thread is already performing the remaining non-incremental compilation steps.
```

Comment:
Our LSP implementation is rather mono thread.
We have 3 threads:

- 1 thread to receive messages and enqueue them
- 1 thread to process messages  from the queue
- 1 thread / timer that just push an event in the queue.

So I'd rather say that everything was designed to allow multithreading, but currently we don't use it. And we have no proof that everything will works fine with multithreading.

Reply:
Rewritten